lib/backend/grpcclient/grpc.go

```diff
@@ -0,0 +1,281 @@
+	return "grpc"
+}
+
+// keep this here to test interface conformance
+var _ backend.Backend = &Backend{}
+
+// New returns new instance of gRPC-powered backend.
+//
+// The supplied context is used to cancel or expire the pending connection to
+// the gRPC server. Once New is returned, the cancellation and expiration of ctx
+// will be noop.
+func New(ctx context.Context, params backend.Params) (*Backend, error) {
+	l := log.WithFields(log.Fields{trace.Component: GetName()})
+
+	l.Infof("Initializing backend %v.", GetName())
+	defer l.Debug("Backend initialization complete.")
+
+	var cfg *Config
+	if err := utils.ObjectToStruct(params, &cfg); err != nil {
+		return nil, trace.BadParameter("gRPC: configuration is invalid: %v", err)
+	}
+
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// CA + client certs for mTLS
+	caCert, err := os.ReadFile(cfg.ServerCA)
+	if err != nil {
+		return nil, trace.ConvertSystemError(err)
+	}
+	caCertPool := x509.NewCertPool()
+	caCertPool.AppendCertsFromPEM(caCert)
+
+	cert, err := tls.LoadX509KeyPair(cfg.ClientCert, cfg.ClientKey)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// TODO(michaelmcallister): There are some tunables that we may need to expose
+	// or set such as:
+	// - InitialConnWindowSize
+	// - InitialWindowSize
+	// - ReadBufferSize
+	// - WriteBuffersize
+	opts := []grpc.DialOption{
+		grpc.WithTransportCredentials(credentials.NewTLS(&tls.Config{
+			MinVersion:   tls.VersionTLS13,
+			RootCAs:      caCertPool,
+			Certificates: []tls.Certificate{cert},
+		})),
+		grpc.WithKeepaliveParams(keepalive.ClientParameters{PermitWithoutStream: true}),
+		grpc.WithBlock(),
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, cfg.DialTimeout)
+	defer cancel()
+	conn, err := grpc.DialContext(ctx, cfg.Server, opts...)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	//TODO(michaelmcallister): Connect to the gRPC server and emit events to the
+	//buffer.
+	buf := backend.NewCircularBuffer(backend.BufferCapacity(cfg.BufferSize))
+	buf.SetInit()
+
+	backend := &Backend{
+		log:        l,
+		clock:      clockwork.NewRealClock(),
+		clientConn: conn,
+		client:     bpb.NewBackendServiceClient(conn),
+		buf:        buf,
+	}
+	return backend, nil
+}
+
+// CheckAndSetDefaults is a helper that returns an error if the supplied
+// configuration is not enough to connect to the gRPC server.
+func (cfg *Config) CheckAndSetDefaults() error {
+	if cfg.Server == "" {
+		return trace.BadParameter("gRPC: server is not specified")
+	}
+	if cfg.ClientCert == "" {
+		return trace.BadParameter("gRPC: client_cert is not specified")
+	}
+	if cfg.ClientKey == "" {
+		return trace.BadParameter("gRPC: client_key is not specified")
+	}
+	if cfg.ServerCA == "" {
+		return trace.BadParameter("gRPC: server_ca is not specified")
+	}
+	if cfg.BufferSize == 0 {
+		cfg.BufferSize = backend.DefaultBufferCapacity
+	}
+	if cfg.DialTimeout == 0 {
+		cfg.DialTimeout = time.Second * 5
```

Comment:
Should this be consistent with the etcd backend?

https://github.com/gravitational/teleport/blob/cc2e1ad214037e999f9085b3c28b311230f9d2ff/lib/backend/etcdbk/etcd.go#L276-L277

https://github.com/gravitational/teleport/blob/cc2e1ad214037e999f9085b3c28b311230f9d2ff/api/defaults/defaults.go#L33

Reply:
Yeah, good idea - thank you!